enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,774 @@
+1. An ostree commit with MicroShift installed is staged.
+1. The sans-MicroShift commit shuts down.
+1. The with-MicroShift commit starts up.
+1. MicroShift starts
+   - No backup action (did not previously run, so there is nothing to backup, and did not previously fail so no restore needed)
+1. MicroShift startup fails
+1. Greenboot runs red scripts
+   - Set backup mode to "restore"
+1. (Failures may need to repeat to trigger rollback.)
+1. The with-MicroShift ostree commit shuts down
+1. The sans-MicroShift ostree commit boots
+1. The agent stages with-MicroShift ostree commit 2
+1. The sans-MicroShift ostree commit shuts down
+1. The with-MicroShift ostree commit 2 starts up.
+1. MicroShift starts
+   - The backup mode is "restore"
+   - There is no backup.
+   - (Open Question) XOR
+     - Remove MicroShift data and start with clean slate
+     - Try running with existing data
+       - What if greenboot fails? Deleting data allows fresh start
+
+**Manual rollback**
+
+<!-- 
+TODO 
+Implies 1 backup per ostree deployment
+-->
+
+**Visual summary**
+
+```mermaid
+flowchart TD
+  start([System boots])
+
+  pre-run[microshift pre-run]
+  run[microshift run]
+
+  exit[Exit with error]
+  red[Red scripts\nNext boot: 'restore']
+  reboot([Greenboot reboots system\nunless it's rollback\notherwise manual intervention is required])
+
+  classDef danger fill:red
+  class exit,red,reboot danger
+
+  green[Green scripts\nNext boot: 'backup']
+  continue-running([System and MicroShift\nproceed to run successfully])
+  classDef good fill:green
+  class continue-running,green good
+  
+%% Actions
+
+  do-backup[Backup data]
+  do-restore[Restore data]
+  do-data-migration[Migrate data]
+  do-persist-version[Persist version metadata]
+  do-clean-data[Remove /var/lib/microshift?\nor try running without cleanup?]
+
+%% Decisions
+
+  data-dir-exists?{/var/lib/microshift/}
+  backup-or-restore?{Backup or restore\nbased on previous boot}
+  backup-exists?{Does backup exist?}
+  version?{Compare\nMicroShift's version:\nmetadata vs binary}
+  data-migration?{Was\nmigration\nsuccessful?}
+  healthy?{"Is MicroShift healthy?\n(greenboot check)"}
+  missing-metadata-version?{Upgrade\nfrom 4.13 is\nsupported}
+  restore-ok?{Restore\nsucceeded?}
+  backup-ok?{Backup\nsucceeded?}
+
+%% Transitions
+
+  start --> pre-run --> data-dir-exists?
+
+  data-dir-exists? -- Doesn't exists --> run
+  data-dir-exists? -- Exists         --> backup-or-restore?
+
+  backup-or-restore? -- "Restore (red)"    --> do-restore
+  backup-or-restore? -- "Backup (green)"   --> do-backup
+  backup-or-restore? -- Missing\nData exists,\nbut no info persisted about what to do\nFits upgrade from 4.13 flow --> do-backup
+
+  do-backup --> backup-ok?
+  backup-ok? -- Yes --> version?
+  backup-ok? -. No .-> exit
+
+  do-restore                             --> backup-exists?
+  backup-exists? -- No --> do-clean-data --> run
+  backup-exists? -- Yes --> restore-ok? 
```

Comment:
(I'm not sure if this is the right place to put this comment.)

It seems from this diagram that on startup the restore is done before the version compatibility is checked. Is that what we want? It seems like we could end up losing data if we're restoring into a state where we will refuse to start because of version incompatibility.

Reply:
So migration and restore would have the same preconditions regarding version skew.
At this point I'm also wondering about "restore" on redboot:
- 2nd deployment, 1st boot
- makes a backup pointing/matching 1st deploy
- red boot, "restore"
- 2nd deployment, 2nd boot
- restore but from what?
Maybe a symlink `/var/lib/microshift.last-backup`...